qtt/simulation/dotsystem.py

```diff
@@ -260,94 +266,131 @@ def __init__(self, name='dotsystem', ndots=3, **kwargs):
+
+    @staticmethod
+    def tunneling_name(dot_left):
+        return 'tun%d' % dot_left
+
+    @staticmethod
+    def tunneling_matrix(dot_left):
+        return DotSystem._matrix_prefix + DotSystem.tunneling_name(dot_left)
+
+    def make_variables(self):
+        """ Create value and matrix for a single variable """
         for name in self.varnames:
-            setattr(self, name, 0)
+            setattr(self, name, 0.)
             # also define that these are float32 numbers!
-            setattr(self, 'M' + name, np.full((self.Nt, self.Nt), 0, dtype=int))
+            setattr(self, self._matrix_prefix + name, np.full((self.number_of_basis_states, self.number_of_basis_states), 0, dtype=int))
 
-    def makevarMs(self, ring=False):
-        ''' Create matrices for the interactions 
+    def _make_variable_matrices(self, ring=False):
+        """ Create matrices for the interactions
+
+        These matrices are used to quickly calculate the Hamiltonian
 
         Args:
-            ring (bool): is the dot array in a ring configuration? (e.g. 2x2)
-        '''
-        m = np.zeros((self.ndots), dtype=int)
+            ring (bool): set to True if the dot array in a ring configuration (e.g. 2x2)
+        """
+        m = np.zeros(self.ndots, dtype=int)
 
         def mkb(i, j):
             mx = m.copy()
             mx[i] = 1
             mx[j] = -1
             return mx
 
-        for i in range(self.Nt):
-            for j in range(self.Nt):
+        potential=range(0,-(self.maxelectrons+1), -1) 
+        on_site_potential = [n*(n-1)/2 for n in range(0, self.maxelectrons+1)]
+        for i in range(self.number_of_basis_states):
+            for j in range(self.number_of_basis_states):
                 if i == j:
                     for dot in range(1, self.ndots + 1):
                         n = self.basis[i, dot - 1]
-                        getattr(self, 'Mdet%i' % dot)[i, i] = [
-                            0, -1, -2, -3][n]  # chemical potential
-                        getattr(self, 'MosC%i' % dot)[i, i] = [
-                            0, 0, 1, 3][n]  # on site charging energy
+                        getattr(self, self.chemical_potential_matrix(dot))[i, i] = potential[n]
+                        getattr(self, self.on_site_charging_matrix(dot))[i, i] = on_site_potential[n]
                         n2 = self.basis[i, dot % self.ndots]
                         n3 = self.basis[i, (dot + 1) % self.ndots]
                         # nearest-neighbour charging energy
-                        getattr(self, 'MisC%i' % dot)[i, i] = n*n2
-                        if hasattr(self, 'isC%i' % (self.ndots + 1)):
+                        logging.info('set inter_site_charging for dot %d' % dot)
+                        getattr(self, self._matrix_prefix + self.inter_site_charging_name(dot))[i, i] = n * n2
+                        
+                        # specific for 2x2 example!!
```

Comment:
@jpdehollain The code from line 349 to 356 is only for the 2x2 sample I believe. The statement `i %2` is related to the state, but I could not figure out what it means. Also the relation between MisC1, ..., MisC6 and the pairs (1,2), (2,3), ... (2,4) is not really clear. The PR now is backwards compatible (I think), but when doing more refactoring I will probably move this to a 2x2 specific section.